agent/consul/state/usage.go

```diff
@@ -75,60 +76,21 @@ func updateUsage(tx WriteTxn, changes Changes) error {
 		case "services":
 			svc := changeObject(change).(*structs.ServiceNode)
 			usageDeltas[change.Table] += delta
-			serviceIter, err := getWithTxn(tx, servicesTableName, "service", svc.ServiceName, &svc.EnterpriseMeta)
-			if err != nil {
-				return err
+
+			// If we have already seen this service name in the transaction changes,
+			// continue on. It is possible for multiple services to be deleted in a
+			// single transaction, via DeleteNode, thus we ensure we do not run the
+			// logic more than once per service-name. Currently this is only a
+			// problem for deletion, since it is not possible for multiple service
+			// instances to be created or updated in the same transaction.
+			if serviceNamesSeen[svc.ServiceName] {
```

Comment:
This will need to consider the namespace, right? Otherwise usage metrics will be wrong when namespaces are used.

Reply:
Ah yes you are right. I feel like I always miss that when constructing these types of sets. Makes me wonder if we could come up with a better way of ensuring that...

Reply:
The `ServiceID` and `ServiceName` types which encompass the name + namespace can be used as map keys without string coercion. We rely on this in the agent local state quite a bit.

Reply:
I think if the fields on the struct were of type `ServiceID` and `ServiceName` (instead of `string`) it would be much easier to do the right thing, and harder to do the wrong thing.

Reply:
I think this is another case of having serializing an internal struct without conversion causing us headaches. When adding namespaces it would have been great to get rid of the free hanging service name/id fields and just hold a single `ServiceName` and `ServiceID`. However that would have altered the serialized format in a backwards incompatible way.

Then there simply wouldn't have been time to refactor all the endpoints to do a conversion much less have machinery in place to catch where the conversion is lossy after updating one struct and not the other.

Reply:
For concrete steps here we could just replace `svc.ServiceName` with `svc.CompoundServiceName()`